### PR TITLE
Add Typesense vector store

### PR DESCRIPTION
**Pull request by jasonbosco**
Related to #931.

Starting this PR to collaborate.

### CODE REVIEW

**Comment by dev2049** [comment]
will try to pick this back up this week!

**Comment by jasonbosco** [comment]
Thank you @dev2049!

File: langchain/vectorstores/typesense.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                                    "name": "vec",
+                                    "type": "float[]",
+                                    "num_dim": len(docs[0]["vec"]),
+                                },
+                                {"name": "text", "type": "string"},
+                                {"name": ".*", "type": "auto"},
+                            ],
+                        }
+                    )
+                except ObjectAlreadyExists:
```
Comment (dev2049) [inline langchain/vectorstores/typesense.py:119]: how could we get an ObjectAlreadyExists if we only go down this path after an ObjectNotFound error?

Comment (jasonbosco) [reply langchain/vectorstores/typesense.py:119]: Hmmm, good point! Added it due to force of habit!

Should be safe to remove the exception handling here.

File: langchain/vectorstores/typesense.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                    self.typesense_client.collections.create(
+                        {
+                            "name": self.typesense_collection_name,
+                            "fields": [
+                                {
+                                    "name": "vec",
+                                    "type": "float[]",
+                                    "num_dim": len(docs[0]["vec"]),
+                                },
+                                {"name": "text", "type": "string"},
```
Comment (dev2049) [inline langchain/vectorstores/typesense.py:114]: is this name supposed to be text_key?

File: langchain/vectorstores/typesense.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        from typesense.exceptions import ObjectAlreadyExists, ObjectNotFound
+
+        # Embed and create the documents
+        docs = []
+        ids = ids or [str(uuid.uuid4()) for _ in texts]
+        for i, text in enumerate(texts):
+            embedding = self._embedding_function(text)
+            metadata = metadatas[i] if metadatas else {}
+            docs.append(
+                {"id": ids[i], "vec": embedding, "text": text, "metadata": metadata}
```
Comment (dev2049) [inline langchain/vectorstores/typesense.py:93]: is "text" supposed to be text_key?

**Comment by dev2049** [comment]
@jasonbosco updated. had a couple outstanding questions and would generally be great to get your eyes on it to make sure the default params make sense. if things generally look ok we can add documentation and tests

File: langchain/vectorstores/typesense.py (hunk 2 above)
Comment (jasonbosco) [reply langchain/vectorstores/typesense.py]: Yup, I've updated it

File: langchain/vectorstores/typesense.py (hunk 3 above)
Comment (jasonbosco) [reply langchain/vectorstores/typesense.py]: Yup, I've updated it

**Comment by jasonbosco** [comment]
@dev2049 Thank you for the updates! I've addressed your comments in the set of commits I just pushed. 

I also found and fixed a few issues I ran into when I used this branch with an [actual langchain example](https://github.com/typesense/langchain-integration-testground).

**Comment by jasonbosco** [comment]
@dev2049 I just added docs for Typesense similar to the other vector stores: https://github.com/jasonbosco/langchain/blob/85b7c739518b928eebb8224fb82fbaf0c3c43010/docs/modules/indexes/vectorstores/examples/typesense.ipynb

**Pull request merged**